### PR TITLE
fix(Pilot Sheet): fix Skill Focus, setup custom reserve retrieval

### LINKED ISSUE #1589: [BUG] skill reserve giving 2 skill points
<strong>Submitted At: </strong>8/3/2021 20:32:08
<strong>Platform: </strong>
Browser - Chrome
<strong>Describe the bug: </strong>
the skill point reserve is supposed give one +2 skill bonus. it is currently giving two
<strong>To Reproduce: </strong>
just add the reserve
<strong>Expected behavior: </strong>
skill point  gives one +2 skill bonus.
<strong>Additional context: </strong>
 version  2.2.28 
<strong>Contact Info: </strong>
Tamaris_Mora_S27#1022 (discord)
<strong>Stack Trace: </strong>

### PR DESCRIPTION
**Pull request by msprijatelj**
# Description

This PR started with fixing an issue introduced by #1564, and developed into realizing we don't actually store bonuses/actions/synergies/etc. for custom-made reserves (i.e., any stemming from Downtime Actions, *NOT* just the 'Custom Reserve' reserve).  Any bonuses, actions, synergies, etc. would only be available for the Reserves listed in `lancer-data`, as reserves with IDs matching those in `lancer-data` are retrieved fresh every time.  This PR aims to address the discrepancy, and hopefully will open the way to a wider variety of custom reserves.

## Issue Number

Closes #1589

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

**Comment by msprijatelj** [comment]
I will freely admit that I am unfamiliar with the proper usage of Serialization/Deserialization and I may be approaching this PR in a completely wrongheaded way.  **Please triple-check me and correct me if this approach is ill-advised.**

File: src/classes/Action.ts [hunk 1]
```diff
@@ -39,9 +39,11 @@ enum ActivePeriod {
 class Frequency {
   public readonly Uses: number
   public readonly Duration: ActivePeriod
+  public readonly FreqText: string
   private _unlimited: boolean
 
   public constructor(frq: string) {
+    this.FreqText = frq
```
Comment (msprijatelj) [inline src/classes/Action.ts:46]: Adding a Serialization function for Frequency felt like overkill, so I stored the original input for the constructor instead. :/

File: src/classes/pilot/Pilot.ts [hunk 2]
```diff
@@ -558,8 +558,7 @@ class Pilot implements ICloudSyncable {
   }
 
   public get MaxSkillPoints(): number {
-    const bonus = this.Reserves.filter(x => x.ID === 'reserve_skill').length
-    return Bonus.IntPilot(Rules.MinimumPilotSkills + this._level + bonus, 'skill_point', this)
+    return Bonus.IntPilot(Rules.MinimumPilotSkills + this._level, 'skill_point', this)
```
Comment (msprijatelj) [inline src/classes/pilot/Pilot.ts:561]: Reverting #1564 to instead add a "bonus" to the Skill Focus reserve.

File: src/ui/components/panels/loadout/active_loadout/dialogs/downtime/_BuySomeTime.vue [hunk 3]
```diff
@@ -120,7 +120,7 @@ export default Vue.extend({
         nr.ResourceCost = 'Only a little time, and only if drastic measures are taken right now'
       else if (this.skillRoll < 20)
         nr.ResourceCost = 'The situation becomes precarious or desperate'
-      this.pilot.Reserves.push(nr)
+      this.pilot.AddReserve(nr)
```
Comment (msprijatelj) [inline src/ui/components/panels/loadout/active_loadout/dialogs/downtime/_BuySomeTime.vue:123]: Noticed that after adding a reserve from Downtime, if I immediately refreshed, the reserve would disappear in many cases.  Replaced all instances of "pushing" a reserve or organization with the appropriate function from Pilot.ts.

File: src/ui/components/panels/loadout/active_loadout/dialogs/downtime/_GetFocused.vue [hunk 4]
```diff
@@ -47,12 +47,16 @@ export default Vue.extend({
...
         new Reserve({
-          id: 'reserve_skill',
+          id: 'reserve_skillfocus',
           type: 'Bonus',
           name: 'Skill Focus',
           description: 'Added via the "Get Focused" Downtime Action',
+          bonuses: [{
+            "id": "skill_point",
+            "val": 1
+          }],
```
Comment (msprijatelj) [inline src/ui/components/panels/loadout/active_loadout/dialogs/downtime/_GetFocused.vue:59]: Following the pattern in [lancer-data/reserves.json](https://github.com/massif-press/lancer-data/blob/master/lib/reserves.json#L1-L11)

File: src/classes/Action.ts (hunk 1 above)
Comment (TranslucentSabre) [reply src/classes/Action.ts:46]: From a consistency standpoint, this isn't the best. However, to be fully consistent we would also need an IFrequencyData struct with only one member which is a little silly.
I think this is fine until and unless we need more than one thing to specify a frequency.

File: src/classes/Bonus.ts [hunk 5]
```diff
@@ -3,7 +3,7 @@ import dict from '@/assets/bonus_dictionary.json'
 
 interface IBonusData {
   id: string
-  val: string | number
+  val: string | number | string[]
```
Comment (TranslucentSabre) [inline src/classes/Bonus.ts:6]: Did you run into an actual issue with this, or did you just notice that this did not match the implementation inside the Bonus class?

File: src/classes/Action.ts (hunk 1 above)
Comment (TranslucentSabre) [reply src/classes/Action.ts:46]: From a consistency standpoint, this isn't the best. However, to be fully consistent we would also need an IFrequencyData struct with only one member which is a little silly.
I think this is fine until and unless we need more than one thing to specify a frequency.

**Comment by TranslucentSabre** [comment]
Overall, I think this is a worthwhile change.

File: src/classes/Bonus.ts (hunk 5 above)
Comment (msprijatelj) [reply src/classes/Bonus.ts:6]: Both; it didn't match the Bonus class implementation, and as a result I couldn't serialize the `Bonus.Value` field into an `IBonusData` object.

File: src/classes/Action.ts (hunk 1 above)
Comment (msprijatelj) [reply src/classes/Action.ts:46]: Understood Re: Consistency; my feelings on this addition align with yours.

**Review by TranslucentSabre — APPROVED** [review]

**Review by jarena3 — APPROVED** [review]

**Pull request merged**